x-pack/plugins/security_solution/public/timelines/components/use_timeline_save_prompt/index.ts

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { i18n } from '@kbn/i18n';
+import { useCallback, useEffect, useMemo } from 'react';
+import { useDispatch } from 'react-redux';
+import type { AppLeaveHandler } from '@kbn/core-application-browser';
+import { useHistory } from 'react-router-dom';
+import type { TimelineId } from '../../../../common/types';
+import { TimelineStatus, TimelineTabs } from '../../../../common/types';
+import { timelineActions } from '../../store/timeline';
+import { useKibana } from '../../../common/lib/kibana';
+import { useDeepEqualSelector } from '../../../common/hooks/use_selector';
+import { getTimelineShowStatusByIdSelector } from '../flyout/selectors';
+import { getLinksWithHiddenTimeline } from '../../../common/links';
+
+// Issue with history.block
+// https://github.com/elastic/kibana/issues/132597
+
+export const useTimelineSavePrompt = (
+  timelineId: TimelineId,
+  onAppLeave: (handler: AppLeaveHandler) => void
+) => {
+  const dispatch = useDispatch();
+
+  const linksWithoutTimelines = getLinksWithHiddenTimeline();
+
+  const pathsWithoutTimelines = linksWithoutTimelines.map((link) => `/app/security${link.path}`);
+
+  const { overlays, application } = useKibana().services;
+  const history = useHistory();
+
+  const prompt = i18n.translate('xpack.securitySolution.timeline.unsavedWorkMessage', {
+    defaultMessage: 'Leave Timeline with unsaved work?',
+  });
+
+  const title = i18n.translate('xpack.securitySolution.timeline.unsavedWorkTitle', {
+    defaultMessage: 'Unsaved changes',
+  });
+
+  const getTimelineShowStatus = useMemo(() => getTimelineShowStatusByIdSelector(), []);
+  const { status: timelineStatus, updated } = useDeepEqualSelector((state) =>
+    getTimelineShowStatus(state, timelineId)
+  );
+
+  const showSaveTimelineModal = useCallback(() => {
+    dispatch(timelineActions.showTimeline({ id: timelineId, show: true }));
+    dispatch(
+      timelineActions.setActiveTabTimeline({
+        id: timelineId,
+        activeTab: TimelineTabs.query,
+      })
+    );
+    dispatch(
+      timelineActions.toggleModalSaveTimeline({
+        id: timelineId,
+        showModalSaveTimeline: true,
+      })
+    );
+  }, [dispatch, timelineId]);
+
+  useEffect(() => {
+    const unblock = history.block((location) => {
+      async function confirmSaveTimeline() {
+        const confirmRes = await overlays?.openConfirm(prompt, {
+          title,
+          'data-test-subj': 'appLeaveConfirmModal',
+        });
+
+        if (confirmRes) {
+          unblock();
+
+          application.navigateToUrl(location.pathname + location.hash + location.search, {
+            state: location.state,
+          });
```

Comment:
calling `location.retry()` does not work?

Reply:
Thanks @semd  for the review.

actually location.retry does not work because, that functionality is available in `history.v5` where [transaction](https://github.com/remix-run/history/blob/dev/docs/blocking-transitions.md) is passed as a param to the `block` callback rather than `location & action` . We are currently on history v4.

For your reference : https://gist.github.com/StringEpsilon/47820cc961c8d82f3b71dc856b5cc616#blocking-transitions